### PR TITLE
Optimisations

### PR DESCRIPTION
**Pull request by je-cook**
A few optimisations to use numpy loops instead of python loops and some typing

### CODE REVIEW

File: src/pyvmcon/vmcon.py [hunk 1]
```diff
@@ -307,18 +304,16 @@ def convergence_value(
         The Lagrange multipliers for inequality constraints for the jth
         evaluation point.
     """
+    ind_eq = min(lamda_equality.shape[0], result.eq.shape[0])
+    ind_ieq = min(lamda_inequality.shape[0], result.ie.shape[0])
```
Comment (je-cook) [inline src/pyvmcon/vmcon.py:308]: This might be overkill but it will ensure the same functionality as zip

File: src/pyvmcon/vmcon.py [hunk 2]
```diff
@@ -402,18 +395,16 @@ def _derivative_lagrangian(
...
-        [lamda * dc for lamda, dc in zip(lamda_equality, result.deq)]
-    )
-    c_inequality_prime = sum(
-        [lamda * dc for lamda, dc in zip(lamda_inequality, result.die)]
-    )
+) -> np.ndarray:
+    ind_eq = min(lamda_equality.shape[0], result.deq.shape[0])
+    ind_ieq = min(lamda_inequality.shape[0], result.die.shape[0])
+    c_equality_prime = (lamda_equality[:ind_eq] * result.deq[:ind_eq]).sum(axis=0)
+    c_inequality_prime = (lamda_inequality[:ind_ieq] * result.die[:ind_ieq]).sum(axis=0)
```
Comment (je-cook) [inline src/pyvmcon/vmcon.py]: The tests fail because of this change with this for me I can't immediately see why:
image

Comment (je-cook) [reply src/pyvmcon/vmcon.py]: scatch that got it working

**Comment by timothy-nunn** [comment]
@je-cook is this ready for review? If not, can you request review when ready?

**Comment by je-cook** [comment]
@timothy-nunn I would but I obviously dont have permissions to do that

**Comment by timothy-nunn** [comment]
@je-cook oh sorry. I assumed everyone in the UKAEA org would be allowed to request review (as is the case on PROCESS?). I'll do that for you tomorrow

**Comment by je-cook** [comment]
Im not in the org so that would be why, I think on PROCESS I might be an external collaborator

**Review by timothy-nunn — APPROVED** [review]
all looks good, thanks for this. Interested to see how much this speeds codes (e.g. PROCESS) up because I did notice that old VMCON vs this VMCON was slightly faster.

**Pull request merged**